File: summary/ruby/2_merge_sort.rb
def merge_sort(arr)
  return arr if arr.length <= 1

  mid   = arr.length / 2
  left  = merge_sort(arr[0...mid])
  right = merge_sort(arr[mid..])

  merge(left, right)
end

def merge(left, right)
  result = []

  until left.empty? || right.empty?
    result << if left.first <= right.first
                left.shift
              else
                right.shift
              end
  end

  result + left + right
end

arr = gets.chomp.split.map(&:to_i)
sorted_arr = merge_sort(arr)

puts "#{sorted_arr}"

File: summary/ts/2_merge_sort.ts
function mergeSort<T>(array: T[]): T[] {
  if (array.length <= 1) {
    return array;
  }

  const middle = Math.floor(array.length / 2);
  const left = array.slice(0, middle);
  const right = array.slice(middle);

  return merge(mergeSort(left), mergeSort(right));
}

function merge<T>(left: T[], right: T[]): T[] {
  let result: T[] = [];
  let leftIndex = 0;
  let rightIndex = 0;

  while (leftIndex < left.length && rightIndex < right.length) {
    if (left[leftIndex] < right[rightIndex]) {
      result.push(left[leftIndex]);
      leftIndex++;
    } else {
      result.push(right[rightIndex]);
      rightIndex++;
    }
  }

  return result.concat(left.slice(leftIndex), right.slice(rightIndex));
}

// 使用例
const numbers = [5, 3, 2, 8, 1, 4];
const sortedNumbers = mergeSort(numbers);
console.log(sortedNumbers); // [1, 2, 3, 4, 5, 8]
